blk/components/virt.c

```diff
@@ -0,0 +1,358 @@
+#include <microkit.h>
+#include <stdint.h>
+#include <stdbool.h>
+#include <sddf/blk/fsmalloc.h>
+#include <sddf/blk/queue.h>
+#include <sddf/blk/mbr.h>
+#include <sddf/blk/util.h>
+#include <sddf/util/cache.h>
+#include <sddf/util/ialloc.h>
+#include <sddf/util/printf.h>
+#include <sddf/util/util.h>
+
+// #define DEBUG_BLK_VIRT
+
+#if defined(DEBUG_BLK_VIRT)
+#define LOG_BLK_VIRT(...) do{ sddf_dprintf("BLK_VIRT|INFO: "); sddf_dprintf(__VA_ARGS__); }while(0)
+#else
+#define LOG_BLK_VIRT(...) do{}while(0)
+#endif
+
+
+#define LOG_BLK_VIRT_ERR(...) do{ sddf_dprintf("BLK_VIRT|ERROR: "); sddf_dprintf(__VA_ARGS__); }while(0)
+
+/* TODO: Currently only works for 1 and 2 clients, need to handle multiple clients */
+
+#define MAX_BLK_NUM_CLIENTS 16
+
+#define DRIVER_CH 1
+#define CLIENT_CH_1 3
+#define CLIENT_CH_2 4
+
+#define MEM_REGION_SIZE 0x200000 //@ericc: autogen this from microkit xml system file
+#define DRV_MAX_DATA_BUFFERS (MEM_REGION_SIZE / BLK_TRANSFER_SIZE)
+
+#define REQBK_SIZE (BLK_NUM_CLIENTS * BLK_QUEUE_SIZE)
+
+uintptr_t blk_config_driver;
+uintptr_t blk_req_queue_driver;
+uintptr_t blk_resp_queue_driver;
+uintptr_t blk_data_driver;
+
+uintptr_t blk_config;
+uintptr_t blk_config2;
+uintptr_t blk_req_queue;
+uintptr_t blk_req_queue2;
+uintptr_t blk_resp_queue;
+uintptr_t blk_resp_queue2;
+uintptr_t blk_data;
+uintptr_t blk_data2;
+
+/* Client specific info */
+typedef struct client {
+    blk_queue_handle_t queue_h;
+    uint32_t ch;
+    uint32_t start_sector;
+    uint32_t sectors;
+} client_t;
+client_t clients[MAX_BLK_NUM_CLIENTS];
+
+blk_queue_handle_t drv_h;
+
+/* Fixed size memory allocator */
+static fsmalloc_t fsmalloc;
+static bitarray_t fsmalloc_avail_bitarr;
+static word_t fsmalloc_avail_bitarr_words[roundup_bits2words64(DRV_MAX_DATA_BUFFERS)];
+
+/* Bookkeeping struct per request */
+typedef struct reqbk {
+    uint32_t cli_id;
+    uint32_t cli_req_id;
+    uintptr_t cli_addr;
+    uintptr_t drv_addr;
+    uint16_t count;
+    blk_request_code_t code;
+} reqbk_t;
+static reqbk_t reqbk[REQBK_SIZE];
+
+/* Index allocator for request bookkeep */
+static ialloc_t ialloc;
+static uint64_t ialloc_idxlist[REQBK_SIZE];
+
+/* Master boot record */
+struct mbr mbr;
+
+bool initialised = false;
```

Comment:
```suggestion
/* The virtualiser is not initialised until we can first read the block device configuration and handle populate the MBR. */
bool initialised = false;
```

Not sure if this is the best phrasing but something like this would be good.